### PR TITLE
Initial Signet support

### PR DESCRIPTION
**Pull request by pavel-main**
## Description

Initial PR to support default Signet ("Signet Global Test Net III"). Custom Signet is not implemented in this PR.

More info about [Signet](https://en.bitcoin.it/wiki/Signet):

- [BIP-0325](https://github.com/bitcoin/bips/blob/master/bip-0325.mediawiki)
- https://github.com/bitcoin/bitcoin/pull/16411
- https://github.com/rust-bitcoin/rust-bitcoin/pull/291

# TODO

- [x] New network parameters
- [x] Bech32 human-readable part changes
- [x] New configuration option
- [x] Genesis block creation
- [ ] Tests to verify genesis block hash

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala [hunk 1]
```diff
@@ -104,6 +100,52 @@ sealed abstract class ChainParams {
...
+      Seq(input),
+      Seq(output),
+      TransactionConstants.lockTime)
+    val prevBlockHash = DoubleSha256Digest(
+      "0000000000000000000000000000000000000000000000000000000000000000")
+    val merkleRootHash = Merkle.computeMerkleRoot(Seq(tx))
+    val genesisBlockHeader =
+      BlockHeader(version, prevBlockHash, merkleRootHash, time, nBits, nonce)
+    val genesisBlock = Block(genesisBlockHeader, Seq(tx))
+    genesisBlock
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala:146]: nit: you could not define this as a `val` and just have `Block(genesisBlockHeader, Seq(tx))`

**Review by benthecarman — APPROVED** [review]

File: core/src/main/scala/org/bitcoins/core/config/NetworkParameters.scala [hunk 2]
```diff
@@ -143,10 +143,43 @@ sealed abstract class RegTest extends BitcoinNetwork {
...
+   */
+  override def rpcPort = 38332
+
+  /**
+   * @inheritdoc
+   */
+  override def dnsSeeds = {
+    Seq(
+      "178.128.221.177",
+      "2a01:7c8:d005:390::5",
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/config/NetworkParameters.scala]: i'm not sure how well supported ipv6 is in our code base and also onion addresses  -- I haven't tried them before. Just an FYI.

File: core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala [hunk 3]
```diff
@@ -104,6 +100,52 @@ sealed abstract class ChainParams {
...
+      amount: CurrencyUnit): Block = {
+    val defaultSignetKey = DoubleSha256Digest(
+      "03ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430")
+    val const = ScriptConstant(defaultSignetKey.bytes)
+
+    val asm = {
+      List(OP_1, const, OP_1, OP_CHECKMULTISIG)
+    }
+
+    val blockScript = ScriptSignature.fromAsm(asm)
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala:131]: i'm not sure how signet's signing works, but you might want to use our built in script types.

You can create a multisig script with

```scala
scala> import org.bitcoins.core.protocol.script._
import org.bitcoins.core.protocol.script._

scala> import org.bitcoins.core.crypto._
import org.bitcoins.core.crypto._

scala> val pubKey = ECPublicKey.fromHex("03ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430")
pubKey: org.bitcoins.core.crypto.ECPublicKey = ECPublicKey(03ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430)

scala> val m = MultiSignatureScriptPubKey(1,Vector(pubKey))
m: org.bitcoins.core.protocol.script.MultiSignatureScriptPubKey = MultiSignatureScriptPubKey(1, 1, ECPublicKey(03ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430))

scala> m.asm
res0: Seq[org.bitcoins.core.script.constant.ScriptToken] = Vector(OP_1, BytesToPushOntoStackImpl(33), ScriptConstantImpl(ByteVector(33 bytes, 0x03ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430)), OP_1, OP_CHECKMULTISIG)
```

This also makes sure you get the asm format correct :-).

Depending on how signet works (raw multisig, vs wrapped in p2wsh/p2sh) you can wrap that script like this:

```scala
scala> val p2sh = P2SHScriptPubKey(m)
p2sh: org.bitcoins.core.protocol.script.P2SHScriptPubKey = P2SHScriptPubKey(Sha256Hash160DigestImpl(04699beafffcd5b24d48e5b5a9f02979aa92d2c0))

scala> val p2wsh = P2WSHWitnessSPKV0(m)
p2wsh: org.bitcoins.core.protocol.script.P2WSHWitnessSPKV0 = wsh(Sha256DigestImpl(d64e78dc1940a14c251ab244ee7d54999018e72894ae85d8d7baf710be6215b6))

```

File: core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala [hunk 4]
```diff
@@ -399,6 +438,57 @@ object RegTestNetChainParams extends BitcoinChainParams {
...
+   * Regtest allows trivial difficulty blocks
+   * [[https://github.com/bitcoin/bitcoin/blob/006d9ab9e59a00fa02b8d5f32919f9d57567fa65/src/chainparams.cpp#L312 regtest min difficulty]]
+   */
+  override lazy val allowMinDifficultyBlocks: Boolean = false
+
+  /**
+   * Regtest allows pow retargetting
+   * [[https://github.com/bitcoin/bitcoin/blob/006d9ab9e59a00fa02b8d5f32919f9d57567fa65/src/chainparams.cpp#L313 signet pow retargetting]]
+   */
+  override lazy val noRetargeting: Boolean = false
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala:484]: Interesting, i'm kind of surprised this is not `true`. So signet allows for POW retargetting? I thought signet got rid of POW all together. cc @kallewoof

File: core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala (hunk 3 above)
Comment (kallewoof) [reply core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala:131]: Signet is raw, I believe; no wrappery. (In fact, I don't think it supports segwit right now. Probably worth looking at, but not hugely important.)

File: core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala (hunk 4 above)
Comment (kallewoof) [reply core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala:484]: Signet in fact includes proof of work! This is, generally, kept at a very low difficulty, however (even lower than main net's lowest difficulty).

**Comment by Christewart** [comment]
Re: Ci failures:

https://travis-ci.org/bitcoin-s/bitcoin-s/jobs/649678802?utm_medium=notification&utm_source=github_status

That should be able to be fixed by `sbt test:scalafmt`

https://travis-ci.org/bitcoin-s/bitcoin-s/jobs/649678795#L2362

^ this CI failure is because we maintain a high test coverage requirement for our `core` module to make sure we don't regress our test suite over time. You can take a look at the coverage reports here: 

https://coveralls.io/builds/28696651

**Pull request closed without merging**